### ISSUE TITLE
`ArchiveManager::archive()` doesn't appear to preserve file timestamps

### ISSUE THREAD

**Issue by thomasfw** [composer/composer#11342]
I've reported an issue in composer/satis (https://github.com/composer/satis/issues/715), but as far as I can tell the issue originates here. Composer's [ArchiveManager::archive()](https://github.com/composer/composer/blob/ecb69eaa4aec3c2bd866d17ee7a46e684dc16f71/src/Composer/Package/Archiver/ArchiveManager.php#L146) doesn't seem to preserve the file modification timestamps from the source when creating an archive.

In Satis, when a package is re-archived, the unzipped files have a modified time of `1 January 1980 at 00:00`. Its important (for caching purposes at least) that the file modified timestamps match the timestamp of the tag/release.

**Comment by stof** [comment]
Zip archives are storing the file modification time. However, the ZipArchiver implemented in Composer does not use [`ZipArchive::setMTimeName`](https://www.php.net/manual/en/ziparchive.setmtimename.php) to set a meaningful value for that timestamp for each file it adds in the archive.
This should be an easy fix though.

**Comment by stof** [comment]
Be careful that `setMtimeName` is defined conditionally in PHP, so this must be guarded by a `method_exists` check (we already do something similar when setting the unix permissions)

**Comment by thomasfw** [comment]
`ZipArchive::setMtimeName()` is only available in php >= 8.0.0 too, whilst the current version of Composer supports >= 7.2.5 Would be good to fix this for all supported php versions, if that's even possible.

**Comment by stof** [comment]
Well, if the method is not available, there is no way to set the mtime in the zip AFAIK.

However, note that using this method will imply using PHP 8+ for the runtime used to generate the archives (so in that case, for the runtime used to run Satis), not for places using composer to install that archive later.

The guard we need to had for older libzip versions will also cover the case of older PHP versions. So to me, that's fine.

**Comment by thomasfw** [comment]
As it turns out, I'm fairly sure the server we were running satis on was missing the zip extension. Since doing a fresh install on php 8.2 (with zip installed), the archived packages seem to have the correct mtime. Presumably the addition of `ZipArchive::setMtimeName()` here would still be relevant though.

This all seems a bit flaky to me. I assume there are multiple zip/unzip fallbacks depending on the runtime, with some being more ideal than others? Also, I realise that satis has separate maintainers, but if composer (and therefore satis) can't create accurate archives on older PHP versions then the minimum required php version for satis should be 8+, and `ext-zip` should be explicitly declared as a dependency if its no longer included by default. There's also a complete lack of documentation available in the satis repo and on composer's website which has made this a nightmare to debug.

If its known that some libraries aren't able to reliably zip/unzip packages (as in, if the archive or unarchived files aren't an accurate copy of the source), maybe composer ought to output warnings when one is being used (related: https://github.com/composer/composer/issues/10535, https://github.com/composer/satis/issues/577).  And maybe [`ZipArchiver`](https://github.com/composer/composer/blob/main/src/Composer/Package/Archiver/ZipArchiver.php) should run some checks on generated archives to compare permissions, timestamps, etc. with the source.

**Comment by stof** [comment]
Well, I don't think it makes sense to add such checks in ZipArchiver: if we have the code to properly generate the archive, the check will never fail. and if we don't have the code in place, we would not have the checks either.

**Comment by thomasfw** [comment]
True, but it looks like setting mtime for files using `ZipArchive` in php < 8 is not possible at all, and so zip/7z would need to be available to reliably create archives. A warning might be appropriate in this case?

I also see that composer docs do [mention possible issues](https://getcomposer.org/doc/articles/troubleshooting.md#zip-archives-are-not-unpacked-correctly-) when relying on php to unpack archives, but no such warnings exist in the satis project when it comes to generating archives. I’m going to update the issue there to hopefully get the docs updated at least, but it seems the maintainers aren’t particularly active.

**Comment by thomasfw** [comment]
True, but it looks like setting mtime for files using `ZipArchive` in php < 8 is not possible at all, and so zip/7z would need to be available to reliably create archives. A warning might be appropriate in this case?

I also see that composer docs do [mention possible issues](https://getcomposer.org/doc/articles/troubleshooting.md#zip-archives-are-not-unpacked-correctly-) when relying on php to unpack archives, but no such warnings exist in the satis project when it comes to generating archives. I’m going to update the issue there to hopefully get the docs updated at least, but it seems the maintainers aren’t particularly active.

**Comment by stof** [comment]
Composer is able to use zip/7z to unzip archives. But the archive _generation_ does not use them.

**Comment by thomasfw** [comment]
In that case, would it not be possible for composer to also use zip/7zip to generate archives when they're available? I imagine that might go some way in fixing this issue, especially on older versions of php where an archive might not be created accurately and `ZipArchive::setMtimeName()` won't be available (i.e. anything prior to php 8).

If that is possible, and assuming composer would fall back to php's `ZipArchive` when zip/7zip aren't available, it might then be appropriate to output a warning of the known issue if the php/libzip versions aren't satisfactory (as is done when [using a version of 7zip with known issues](https://github.com/composer/composer/commit/30671fb9040a06d2369ade29e8fcae3b380c9b8e) to unpack an archive).

**Comment by stof** [comment]
I fear that implement an archiver would be a lot more complex than implementing the installer.

However, it is also much easier to control the PHP runtime of the server where you run Satis (or an other tool generating composer archives) than to control the PHP runtime when the installer runs.
So to me, fixing the ZipArchiver to use `ZipArchive::setMtimeName()` is enough (and it will probably take less of your time to update PHP on your build server generating the Satis archive than it would take to maintain this archiver based on the zip CLI)

**Comment by leagris** [comment]
ZipARchive in php7.4 do preserve mtime of files, but cannot from directories if you use the addEmptyDirectory method.
Although, there is a workaround I found using the standard ext-zip ZipArchive methods with the help a a temporary empty file used to create directory entries with proper mtime in the zip archive. See: https://stackoverflow.com/a/78407950/7939871
